Port SerializedScriptValue to the new IPC serialization format

cdumez opened the pull request:
#### 197177abbd7c527be4856a691f8f42c6c1768f8d
<pre>
Port SerializedScriptValue to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=265904">https://bugs.webkit.org/show_bug.cgi?id=265904</a>

Reviewed by Alex Christensen.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBufferContents::fromDataSpan):
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::SerializedScriptValue::SerializedScriptValue):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
(WebCore::SerializedScriptValue::arrayBufferContentsArray const):
(WebCore::SerializedScriptValue::detachedRTCDataChannels const):
(WebCore::SerializedScriptValue::serializedVideoChunks const):
(WebCore::SerializedScriptValue::serializedAudioChunks const):
(WebCore::SerializedScriptValue::encode const): Deleted.
(WebCore::SerializedScriptValue::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::SerializedScriptValue&gt;&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::SerializedScriptValue&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/271671@main">https://commits.webkit.org/271671@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/795e53c8e90e2dd24208ffeabbaaa225448859c9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29173 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30509 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31760 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26532 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29769 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9989 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26536 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29445 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6505 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24995 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5609 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5754 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26031 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33097 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/25163 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26649 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26465 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31979 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/29483 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5715 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3895 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29764 "Found 2 new API test failures: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/close-quickly, /WebKitGTK/TestWebKitUserContentFilterStore:/webkit/WebKitUserContentFilterStore/save-from-file (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7392 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/35778 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/6967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6226 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7714 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3751 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6237 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -6782,6 +6787,30 @@ enum class WebCore::InspectorClientDeveloperPreference : uint8_t {
+#endif
+
+[RefCounted, CreateUsing=createFromIPC] class WebCore::SerializedScriptValue {
+    Vector<uint8_t> wireBytes();
```
It would be nice if we could eventually make this even more declarative.

achristensen07 commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -652,5 +652,17 @@ ASCIILiteral errorMessageForTransfer(ArrayBuffer* buffer)
+    if (!buffer)
+        return std::nullopt;
+
+    static_assert(sizeof(std::span<const uint8_t>::element_type) == 1);
```
I don't think this is necessary.  If sizeof(uint8_t) isn't 1, you're going to have big problems.

achristensen07 commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -5465,25 +5474,25 @@ JSValue SerializedScriptValue::deserialize(JSGlobalObject& lexicalGlobalObject,
 JSValue SerializedScriptValue::deserialize(JSGlobalObject& lexicalGlobalObject, JSGlobalObject* globalObject, const Vector<RefPtr<MessagePort>>& messagePorts, const Vector<String>& blobURLs, const Vector<String>& blobFilePaths, SerializationErrorMode throwExceptions, bool* didFail)
 {
-    DeserializationResult result = CloneDeserializer::deserialize(&lexicalGlobalObject, globalObject, messagePorts, WTFMove(m_backingStores)
+    DeserializationResult result = CloneDeserializer::deserialize(&lexicalGlobalObject, globalObject, messagePorts, WTFMove(m_internals.backingStores)
```
Can we just pass WTFMove(m_internals) to CloneDeserializer::deserialize instead of all these parameter pieces of it?

cdumez replied:
Not easily. I'll try looking in a follow-up as trying to make this change ended up too large.

cdumez committed:
Port SerializedScriptValue to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=265904

Reviewed by Alex Christensen.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBufferContents::fromDataSpan):
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::SerializedScriptValue::SerializedScriptValue):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
(WebCore::SerializedScriptValue::arrayBufferContentsArray const):
(WebCore::SerializedScriptValue::detachedRTCDataChannels const):
(WebCore::SerializedScriptValue::serializedVideoChunks const):
(WebCore::SerializedScriptValue::serializedAudioChunks const):
(WebCore::SerializedScriptValue::encode const): Deleted.
(WebCore::SerializedScriptValue::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<RefPtr<WebCore::SerializedScriptValue>>::encode): Deleted.
(IPC::ArgumentCoder<RefPtr<WebCore::SerializedScriptValue>>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/271671@main
